File: sq/core/math/Vec2.ts

/**
 * 二维向量
 * */

import { sqclass, type, serializable, SQFloat } from "../data/index";
import { CommonUtils } from "../utils";

@sqclass("sq.Vec2")
export default class Vec2 {


    public static temp1Vec2 = new Vec2();
    public static temp2Vec2 = new Vec2();

    @type(SQFloat)
    @serializable
    public x: number;

    @type(SQFloat)
    @serializable
    public y: number;

    constructor(x?: number, y?: number) {
        this.x = x || 0;
        this.y = y || 0;
    }

    /**
     * 向量加法
     * @param rhs
     */
    add(rhs: Vec2) {
        this.x += rhs.x;
        this.y += rhs.y;
        return this;
    }

    add2(lhs: Vec2, rhs: Vec2) {
        this.x = lhs.x + rhs.x;
        this.y = lhs.y + rhs.y;
        return this;
    }

    sub(rhs: Vec2) {
        this.x -= rhs.x;
        this.y -= rhs.y;
        return this;
    }


    sub2(lhs: Vec2, rhs: Vec2) {
        this.x = lhs.x - rhs.x;
        this.y = lhs.y - rhs.y;
        return this;
    }

    scale(scalar: number) {
        this.x *= scalar;
        this.y *= scalar;

        return this;
    }

    mul(rhs: Vec2) {
        this.x *= rhs.x;
        this.y *= rhs.y;
        return this;
    }

    mul2(lhs: Vec2, rhs: Vec2) {
        this.x = lhs.x * rhs.x;
        this.y = lhs.x * rhs.y;
        return this;
    }

    dot(rhs: Vec2) {
        return this.x * rhs.x + this.y * rhs.y;
    }

    cross(rhs: Vec2) {
        return this.x * rhs.y - this.y * rhs.x;
    }

    crossSV(s: number) {
        let x = this.x;
        this.x = this.y * -s;
        this.y = x * s;
        return this;
    }

    abs() {
        this.x = Math.abs(this.x);
        this.y = Math.abs(this.y);
        return this;
    }

    equals(rhs: Vec2): boolean {
        return this.x === rhs.x && this.y === rhs.y;
    }

    length(): number {
        return Math.sqrt(this.x * this.x + this.y * this.y);
    }

    lengthSq() {
        return this.x * this.x + this.y * this.y;
    }

    lerp(lhs: Vec2, rhs: Vec2, alpha: number) {
        this.x = lhs.x + alpha * (rhs.x - lhs.x);
        this.y = lhs.y + alpha * (rhs.y - lhs.y);
        return this;
    }


    normalize() {
        var lengthSq = this.lengthSq();
        if (lengthSq > 0) {
            var invLength = 1 / Math.sqrt(lengthSq);
            this.x *= invLength;
            this.y *= invLength;
        }
        return this;
    }


    set(x: number, y: number) {
        this.x = x;
        this.y = y;
        return this;
    }

    copy(vec: Vec2) {
        this.x = vec.x;
        this.y = vec.y;
        return this;
    }

    toString() {
        return "(" + this.x + "," + this.y + ")";
    }

    static cross(rv: Vec2, lv: Vec2): number {
        return rv.x * lv.y - lv.x * rv.y;
    }

    static dot(rv: Vec2, lv: Vec2): number {
        return rv.x * lv.x + rv.y * rv.y;
    }

    static min(vec1: Vec2, vec2: Vec2, result: Vec2): Vec2 {
        result.x = Math.min(vec1.x, vec2.x);
        result.y = Math.min(vec2.y, vec2.y);
        return result;
    }

    static max(vec1: Vec2, vec2: Vec2, result: Vec2): Vec2 {
        result.x = Math.max(vec1.x, vec2.x);
        result.y = Math.max(vec1.y, vec2.y);
        return result;
    }

    static isValid(a: Vec2): boolean {
        return CommonUtils.isDefine(a.x) && CommonUtils.isDefine(a.y);
    }

    static mulAdd(a: Vec2, s: number, b: Vec2, result?: Vec2): Vec2 {
        result = result || new Vec2();
        result.x = a.x + s * b.x;
        result.y = a.y + s * b.y
        return result;
    }

    static distanceSquared(a: Vec2, b: Vec2): number {
        let x: number = b.x - a.x;
        let y: number = b.y - a.y;
        return x * x + y * y;
    }

    static leftPerp(v: Vec2, out: Vec2) : Vec2 {
        let x = -v.y;
        out.y = v.x;
        out.x = x;
        return out;
    }

    static rightPerp(v: Vec2, out: Vec2): Vec2 {
        let x = v.y;
        out.y = -v.x;
        out.x = x;
        return out;
    }
}

File: native/engine/core/math/Vec2.cpp
#include "Vec2.h"
#include <cmath>
#include "math.h"
#include <stdio.h>

Vec2::Vec2(float x, float y) : x(x), y(y) {}
Vec2::Vec2() : x(0), y(0) {}
Vec2::Vec2(float v) : x(v), y(v) {}
Vec2::Vec2(const Vec2 &v)
{
    // printf("++++copy\n");
    x = v.x;
    y = v.y;
}
bool Vec2::equals(const Vec2 &v)
{
    return this->x == v.x && this->y == v.y;
}

void Vec2::set(float x, float y)
{
    this->x = x;
    this->y = y;
};

void Vec2::add(const Vec2 &rhs)
{
    this->x += rhs.x;
    this->y += rhs.y;
}

void Vec2::add2(const Vec2 &lhs, const Vec2 &rhs)
{
    this->x = lhs.x + rhs.x;
    this->y = lhs.y + rhs.y;
}

void Vec2::sub(const Vec2 &rhs)
{
    this->x -= rhs.x;
    this->y -= rhs.y;
}

void Vec2::sub2(const Vec2 &lhs, const Vec2 &rhs)
{
    this->x = lhs.x - rhs.x;
    this->y = lhs.y - rhs.y;
}

void Vec2::scale(float scalar)
{
    this->x *= scalar;
    this->y *= scalar;
}

void Vec2::mul(const Vec2 &rhs)
{
    this->x *= rhs.x;
    this->y *= rhs.y;
}

void Vec2::mul(float s)
{
    x *= s;
    y *= s;
}

void Vec2::mul2(const Vec2 &lhs, const Vec2 &rhs)
{
    this->x = lhs.x * rhs.x;
    this->y = lhs.x * rhs.y;
}

float Vec2::dot(const Vec2 &rhs)
{
    return this->x * rhs.x + this->y * rhs.y;
}

float Vec2::cross(const Vec2 &rhs)
{
    return this->x * rhs.y - this->y * rhs.x;
}

void Vec2::copy(const Vec2 &vec)
{
    this->x = vec.x;
    this->y = vec.y;
}

float Vec2::lengthSqr() const
{
    return Vec2::lengthSqr(*this);
}

float Vec2::normalize()
{
    return Vec2::normalize(*this, *this);
}

float Vec2::normalize(const Vec2 &a, Vec2 &out)
{
    float x = a.x;
    float y = a.y;
    float len = sqrt(x * x + y * y);
    if (len > 0)
    {
        float invLen = 1 / len;
        out.x = x * invLen;
        out.y = y * invLen;
    }
    else
    {
        out.x = 0;
        out.y = 0;
    }
    return len;
}

float Vec2::len() const
{
    return sqrt(x * x + y * y);
}

float Vec2::lengthSqr(const Vec2 &v)
{
    return v.x * v.x + v.y * v.y;
}

void Vec2::skew()
{
    float x1 = x;
    x = -y;
    y = x1;
}

float Vec2::length(const Vec2 &v1, const Vec2 &v2)
{
    Vec2 temp;
    temp.set(v1.x - v2.x, v1.y - v2.y);
    return temp.len();
}

float Vec2::dot(const Vec2 &v1, const Vec2 &v2)
{
    Vec2 temp;
    temp.set(v1.x, v1.y);
    return temp.dot(v2);
}

float Vec2::cross(const Vec2 &v1, const Vec2 &v2)
{
    Vec2 temp;
    temp.set(v1.x, v1.y);
    return temp.cross(v2);
}

void Vec2::lerp(const Vec2 &v1, const Vec2 &v2, float alpha, Vec2 &out)
{
    out.x = Math::lerp(v1.x, v2.x, alpha);
    out.y = Math::lerp(v1.y, v2.y, alpha);
}

void Vec2::abs(Vec2 &v1)
{
    v1.x = std::fabs(v1.x);
    v1.y = std::fabs(v1.y);
}

void Vec2::max(Vec2 &v1, float v)
{
    v1.x = std::fmax(v1.x, v);
    v1.y = std::fmax(v1.y, v);
}

void Vec2::leftPerp(const Vec2 &v, Vec2 &out)
{
    float x = -v.y;
    out.y = v.x;
    out.x = x;
}

void Vec2::rightPerp(const Vec2 &v, Vec2 &out)
{
    float x = v.y;
    out.y = -v.x;
    out.x = x;
}

Vec2 Vec2::operator*(const Vec2 &v1) const
{
    return Vec2(x * v1.x, y * v1.y);
}

Vec2 Vec2::operator*(float v) const
{
    return Vec2(x * v, y * v);
}

Vec2 &Vec2::operator*=(float v)
{
    x *= v;
    y *= v;
    return *this;
}

Vec2 Vec2::operator+(const Vec2 &v1) const
{
    return Vec2(x + v1.x, y + v1.y);
}

Vec2 Vec2::operator+(float v) const
{
    return Vec2(x + v, y + v);
}

Vec2 &Vec2::operator+=(float v)
{
    x += v;
    y += v;
    return *this;
}

Vec2 &Vec2::operator+=(const Vec2 &v)
{
    x += v.x;
    y += v.y;
    return *this;
}

bool Vec2::operator==(const Vec2 &c) const
{
    return x == c.x && y == c.y;
}

bool Vec2::operator!=(const Vec2 &c) const
{
    return x != c.x || y != c.y;
}

Vec2 Vec2::operator-(const Vec2 &v1) const
{
    return Vec2(x - v1.x, y - v1.y);
}

Vec2 Vec2::operator/(float v) const
{
    return Vec2(x / v, y / v);
}
